File: scripts/test-fb-permissions.ts
/**
 * Test Facebook Graph API Permissions
 * =====================================
 * 
 * Script này:
 * 1. Lấy token từ database (decrypt nếu cần)
 * 2. Test tất cả API permissions
 * 3. Output kết quả chi tiết
 * 
 * Usage:
 *   npx ts-node scripts/test-fb-permissions.ts
 *   
 * Hoặc (nếu dùng token thủ công):
 *   FB_TOKEN=your_token FB_PAGE_ID=123456 npx ts-node scripts/test-fb-permissions.ts
 */

import { createClient } from '@supabase/supabase-js';
import crypto from 'crypto';
import 'dotenv/config';

const GRAPH_API_VERSION = 'v21.0';
const BASE_URL = `https://graph.facebook.com/${GRAPH_API_VERSION}`;

// Encryption settings (same as tokenManager.ts)
const ENCRYPTION_KEY = process.env.FACEBOOK_TOKEN_ENCRYPTION_KEY || 'default-key-change-in-production-32bytes';
const ALGORITHM = 'aes-256-cbc';

interface TestResult {
  permission: string;
  endpoint: string;
  success: boolean;
  details: string;
}

const results: TestResult[] = [];

// ============================================
// HELPER FUNCTIONS
// ============================================

function decryptToken(encryptedData: string): string {
  try {
    const [ivHex, encrypted] = encryptedData.split(':');
    if (!encrypted) {
      // Not encrypted, return as-is
      return encryptedData;
    }
    
    const iv = Buffer.from(ivHex, 'hex');
    const key = Buffer.from(ENCRYPTION_KEY.padEnd(32, '0').slice(0, 32));
    const decipher = crypto.createDecipheriv(ALGORITHM, key, iv);
    
    let decrypted = decipher.update(encrypted, 'hex', 'utf8');
    decrypted += decipher.final('utf8');
    
    return decrypted;
  } catch {
    // If decryption fails, assume it's not encrypted
    return encryptedData;
  }
}

async function makeRequest(
  token: string,
  method: 'GET' | 'POST' | 'DELETE',
  endpoint: string,
  params: Record<string, any> = {},
  body?: Record<string, any>
): Promise<{ data: any; error: string | null }> {
  const url = new URL(`${BASE_URL}/${endpoint}`);
  url.searchParams.set('access_token', token);
  
  for (const [key, value] of Object.entries(params)) {
    url.searchParams.set(key, String(value));
  }
  
  try {
    const options: RequestInit = {
      method,
      headers: { 'Content-Type': 'application/json' },
    };
    
    if (body && method !== 'GET') {
      options.body = JSON.stringify(body);
    }
    
    const response = await fetch(url.toString(), options);
    const data = await response.json();
    
    if (data.error) {
      return { data: null, error: `${data.error.message} (code: ${data.error.code})` };
    }
    
    return { data, error: null };
  } catch (err) {
    return { data: null, error: String(err) };
  }
}

function logResult(permission: string, endpoint: string, success: boolean, details: string) {
  results.push({ permission, endpoint, success, details });
  
  const status = success ? '✅ PASS' : '❌ FAIL';
  console.log(`${status} | ${permission} | ${endpoint}`);
  if (details) {
    console.log(`       └── ${details.slice(0, 100)}`);
  }
}

// ============================================
// TEST FUNCTIONS FOR EACH PERMISSION
// ============================================

async function testPagesReadUserContent(token: string, pageId: string) {
  console.log('\n' + '='.repeat(60));
  console.log('Testing: pages_read_user_content');
  console.log('='.repeat(60));
  
  // Get feed (includes visitor posts)
  const { data, error } = await makeRequest(token, 'GET', `${pageId}/feed`, {
    fields: 'id,message,from,created_time,type',
    limit: 5
  });
  
  if (error) {
    logResult('pages_read_user_content', `${pageId}/feed`, false, error);
  } else if (data?.data) {
    logResult('pages_read_user_content', `${pageId}/feed`, true, `Found ${data.data.length} posts`);
  } else {
    logResult('pages_read_user_content', `${pageId}/feed`, false, 'No data returned');
  }
}

async function testPagesReadEngagement(token: string, pageId: string) {
  console.log('\n' + '='.repeat(60));
  console.log('Testing: pages_read_engagement');
  console.log('='.repeat(60));
  
  // Get page insights
  const { data, error } = await makeRequest(token, 'GET', `${pageId}/insights`, {
    metric: 'page_engaged_users,page_post_engagements',
    period: 'day'
  });
  
  if (error) {
    logResult('pages_read_engagement', `${pageId}/insights`, false, error);
  } else if (data?.data) {
    logResult('pages_read_engagement', `${pageId}/insights`, true, `Got ${data.data.length} metrics`);
  } else {
    logResult('pages_read_engagement', `${pageId}/insights`, false, 'No data returned');
  }
  
  // Get posts with engagement
  const { data: postsData, error: postsError } = await makeRequest(token, 'GET', `${pageId}/posts`, {
    fields: 'id,message,likes.summary(true),comments.summary(true),shares',
    limit: 5
  });
  
  if (postsError) {
    logResult('pages_read_engagement', `${pageId}/posts (engagement)`, false, postsError);
  } else if (postsData?.data) {
    logResult('pages_read_engagement', `${pageId}/posts (engagement)`, true, 
      `Got engagement for ${postsData.data.length} posts`);
  }
}

async function testPagesManagePosts(token: string, pageId: string) {
  console.log('\n' + '='.repeat(60));
  console.log('Testing: pages_manage_posts');
  console.log('='.repeat(60));
  
  // Create unpublished test post
  const { data, error } = await makeRequest(token, 'POST', `${pageId}/feed`, {}, {
    message: `[API Test] Created at ${new Date().toISOString()}. Will be deleted.`,
    published: false
  });
  
  if (error) {
    logResult('pages_manage_posts', `${pageId}/feed (CREATE)`, false, error);
    return;
  }
  
  if (data?.id) {
    const postId = data.id;
    logResult('pages_manage_posts', `${pageId}/feed (CREATE)`, true, `Created: ${postId}`);
    
    // Delete the test post
    const { data: delData, error: delError } = await makeRequest(token, 'DELETE', postId);
    
    if (delError) {
      logResult('pages_manage_posts', `${postId} (DELETE)`, false, delError);
    } else if (delData?.success) {
      logResult('pages_manage_posts', `${postId} (DELETE)`, true, 'Deleted successfully');
    }
  } else {
    logResult('pages_manage_posts', `${pageId}/feed (CREATE)`, false, 'No post ID returned');
  }
}

async function testPagesManageEngagement(token: string, pageId: string) {
  console.log('\n' + '='.repeat(60));
  console.log('Testing: pages_manage_engagement');
  console.log('='.repeat(60));
  
  // Get a recent post
  const { data: postsData } = await makeRequest(token, 'GET', `${pageId}/posts`, {
    fields: 'id,message',
    limit: 1
  });
  
  if (!postsData?.data?.[0]) {
    logResult('pages_manage_engagement', 'GET posts', false, 'No posts found');
    return;
  }
  
  const postId = postsData.data[0].id;
  
  // Get comments
  const { data: commentsData, error } = await makeRequest(token, 'GET', `${postId}/comments`, {
    fields: 'id,message,from,created_time',
    limit: 5
  });
  
  if (error) {
    logResult('pages_manage_engagement', `${postId}/comments (READ)`, false, error);
  } else if (commentsData?.data) {
    logResult('pages_manage_engagement', `${postId}/comments (READ)`, true, 
      `Got ${commentsData.data.length} comments`);
    
    // If there are comments, try to reply
    if (commentsData.data.length > 0) {
      const commentId = commentsData.data[0].id;
      
      const { data: replyData, error: replyError } = await makeRequest(
        token, 'POST', `${commentId}/comments`, {}, 
        { message: '[API Test Reply] Thank you!' }
      );
      
      if (replyError) {
        logResult('pages_manage_engagement', `${commentId}/comments (REPLY)`, false, replyError);
      } else if (replyData?.id) {
        logResult('pages_manage_engagement', `${commentId}/comments (REPLY)`, true, 
          `Created reply: ${replyData.id}`);
        
        // Clean up
        await makeRequest(token, 'DELETE', replyData.id);
        console.log('       └── Test reply cleaned up');
      }
    }
  }
}

async function testBusinessManagement(token: string) {
  console.log('\n' + '='.repeat(60));
  console.log('Testing: business_management');
  console.log('='.repeat(60));
  
  const { data, error } = await makeRequest(token, 'GET', 'me/businesses', {
    fields: 'id,name,created_time'
  });
  
  if (error) {
    logResult('business_management', 'me/businesses', false, error);
  } else if (data?.data) {
    logResult('business_management', 'me/businesses', true, `Found ${data.data.length} businesses`);
    
    if (data.data.length > 0) {
      const businessId = data.data[0].id;
      
      const { data: pagesData, error: pagesError } = await makeRequest(
        token, 'GET', `${businessId}/owned_pages`,
        { fields: 'id,name' }
      );
      
      if (pagesError) {
        logResult('business_management', `${businessId}/owned_pages`, false, pagesError);
      } else if (pagesData?.data) {
        logResult('business_management', `${businessId}/owned_pages`, true, 
          `Found ${pagesData.data.length} pages`);
      }
    }
  }
}

async function testAdditionalPermissions(token: string, pageId: string) {
  console.log('\n' + '='.repeat(60));
  console.log('Testing: Additional permissions (verification)');
  console.log('='.repeat(60));
  
  // pages_show_list
  const { data: accountsData, error: accountsError } = await makeRequest(token, 'GET', 'me/accounts', {
    fields: 'id,name,category'
  });
  
  if (accountsError) {
    logResult('pages_show_list', 'me/accounts', false, accountsError);
  } else if (accountsData?.data) {
    logResult('pages_show_list', 'me/accounts', true, `Found ${accountsData.data.length} pages`);
  }
  
  // pages_manage_metadata
  const { data: pageData, error: pageError } = await makeRequest(token, 'GET', pageId, {
    fields: 'id,name,about,category,phone,website'
  });
  
  if (pageError) {
    logResult('pages_manage_metadata', pageId, false, pageError);
  } else if (pageData?.id) {
    logResult('pages_manage_metadata', pageId, true, `Page: ${pageData.name}`);
  }
}

function printSummary() {
  console.log('\n' + '='.repeat(60));
  console.log('TEST SUMMARY');
  console.log('='.repeat(60));
  
  const passed = results.filter(r => r.success).length;
  const failed = results.filter(r => !r.success).length;
  
  console.log(`\nTotal tests: ${results.length}`);
  console.log(`✅ Passed: ${passed}`);
  console.log(`❌ Failed: ${failed}`);
  
  if (failed > 0) {
    console.log('\nFailed tests:');
    results.filter(r => !r.success).forEach(r => {
      console.log(`  - ${r.permission}: ${r.endpoint}`);
      console.log(`    ${r.details.slice(0, 80)}`);
    });
  }
}

// ============================================
// MAIN
// ============================================

async function main() {
  console.log('='.repeat(60));
  console.log('FACEBOOK GRAPH API PERMISSION TEST');
  console.log('='.repeat(60));
  
  let token = process.env.FB_TOKEN || process.env.FACEBOOK_PAGE_ACCESS_TOKEN;
  let pageId = process.env.FB_PAGE_ID || process.env.FACEBOOK_PAGE_ID;
  
  // If not provided via env, get from database
  if (!token || !pageId) {
    console.log('\nNo token in env, fetching from database...');
    
    const supabase = createClient(
      process.env.NEXT_PUBLIC_SUPABASE_URL!,
      process.env.SUPABASE_SERVICE_ROLE_KEY!
    );
    
    const { data: pages, error } = await supabase
      .from('facebook_pages')
      .select('*')
      .eq('status', 'active')
      .limit(1);
    
    if (error || !pages?.length) {
      console.error('❌ No active pages found in database');
      console.log('\nUsage:');
      console.log('  FB_TOKEN=<token> FB_PAGE_ID=<page_id> npx ts-node scripts/test-fb-permissions.ts');
      process.exit(1);
    }
    
    const page = pages[0];
    pageId = page.page_id;
    token = decryptToken(page.page_access_token);
    
    console.log(`\nUsing page: ${page.page_name} (${pageId})`);
  }
  
  if (!token || !pageId) {
    console.error('Missing token or page ID after lookup');
    process.exit(1);
  }

  console.log(`\nAPI Version: ${GRAPH_API_VERSION}`);
  console.log(`Page ID: ${pageId}`);
  console.log(`Token: ${token.slice(0, 20)}...`);
  
  // Run all tests
  await testPagesReadUserContent(token, pageId);
  await testPagesReadEngagement(token, pageId);
  await testPagesManagePosts(token, pageId);
  await testPagesManageEngagement(token, pageId);
  await testBusinessManagement(token);
  await testAdditionalPermissions(token, pageId);
  
  printSummary();
}

main().catch(console.error);

File: scripts/test-fb-permissions.py
#!/usr/bin/env python3
"""
Test Facebook Graph API permissions.

Usage:
  FB_TOKEN=... FB_PAGE_ID=... python scripts/test-fb-permissions.py
  (or set FACEBOOK_USER_ACCESS_TOKEN in .env)

Optional:
  FB_GRAPH_API_VERSION=v24.0
  FB_INSIGHTS_METRICS=page_engaged_users,page_post_engagements
  FB_MESSAGING_RECIPIENT_ID=... (optional, for send message test)

If FB_TOKEN/FB_PAGE_ID are not set, the script attempts to load the first
active page from Supabase and uses page_access_token + page_id.
"""

from __future__ import annotations

import json
import os
import sys
import urllib.error
import urllib.parse
import urllib.request
from datetime import datetime
from typing import Any, Dict, List, Optional, Tuple

# Ensure UTF-8 output on Windows terminals
try:
    if hasattr(sys.stdout, 'reconfigure'):
        sys.stdout.reconfigure(encoding='utf-8', errors='replace')
    if hasattr(sys.stderr, 'reconfigure'):
        sys.stderr.reconfigure(encoding='utf-8', errors='replace')
except Exception:
    pass

GRAPH_API_VERSION = os.environ.get("FB_GRAPH_API_VERSION", "v24.0")
BASE_URL = f"https://graph.facebook.com/{GRAPH_API_VERSION}"

REQUIRED_PERMISSIONS = {
    "pages_show_list",
    "pages_read_engagement",
    "pages_manage_posts",
    "pages_messaging",
    "pages_manage_metadata",
}

ENCRYPTION_KEY = os.environ.get(
    "FACEBOOK_TOKEN_ENCRYPTION_KEY",
    "default-key-change-in-production-32bytes",
)

# Optional AES support if pycryptodome is installed
try:
    from Crypto.Cipher import AES  # type: ignore
except Exception:
    AES = None


class TestResult:
    def __init__(self, permission: str, endpoint: str, success: bool, details: str) -> None:
        self.permission = permission
        self.endpoint = endpoint
        self.success = success
        self.details = details


results: List[TestResult] = []
skipped: List[TestResult] = []


def load_env(path: str = ".env") -> None:
    if not os.path.exists(path):
        return
    try:
        with open(path, "r", encoding="utf-8") as handle:
            for line in handle:
                line = line.strip()
                if not line or line.startswith("#"):
                    continue
                if "=" not in line:
                    continue
                key, value = line.split("=", 1)
                key = key.strip()
                value = value.strip().strip('"').strip("'")
                if key and key not in os.environ:
                    os.environ[key] = value
    except Exception:
        return


def decrypt_token(encrypted_data: str) -> str:
    if ":" not in encrypted_data:
        return encrypted_data
    if AES is None:
        raise RuntimeError(
            "Encrypted token detected but pycryptodome is not installed. "
            "Set FB_TOKEN or install pycryptodome."
        )

    try:
        iv_hex, encrypted_hex = encrypted_data.split(":", 1)
        iv = bytes.fromhex(iv_hex)
        encrypted = bytes.fromhex(encrypted_hex)
        key = ENCRYPTION_KEY.ljust(32, "0")[:32].encode("utf-8")
        cipher = AES.new(key, AES.MODE_CBC, iv)
        decrypted = cipher.decrypt(encrypted)
        pad_len = decrypted[-1]
        if pad_len < 1 or pad_len > 16:
            return decrypted.decode("utf-8", errors="ignore")
        return decrypted[:-pad_len].decode("utf-8", errors="ignore")
    except Exception:
        return encrypted_data


def http_request(
    method: str,
    url: str,
    headers: Optional[Dict[str, str]] = None,
    body: Optional[Dict[str, Any]] = None,
) -> Tuple[Optional[Dict[str, Any]], Optional[str]]:
    try:
        data_bytes = None
        final_headers = headers or {}
        if body is not None and method.upper() != "GET":
            data_bytes = json.dumps(body).encode("utf-8")
            final_headers.setdefault("Content-Type", "application/json")

        req = urllib.request.Request(url, data=data_bytes, method=method.upper())
        for key, value in final_headers.items():
            req.add_header(key, value)

        with urllib.request.urlopen(req) as response:
            raw = response.read().decode("utf-8")
            data = json.loads(raw) if raw else {}
            return data, None
    except urllib.error.HTTPError as exc:
        raw = exc.read().decode("utf-8", errors="replace")
        parsed: Optional[Dict[str, Any]] = None
        try:
            parsed = json.loads(raw) if raw else None
        except json.JSONDecodeError:
            parsed = None

        if parsed:
            return parsed, f"HTTP {exc.code} {exc.reason} | Body: {raw}"
        return None, f"HTTP {exc.code} {exc.reason} | Body: {raw}"
    except Exception as exc:
        return None, str(exc)


def make_request(
    token: str,
    method: str,
    endpoint: str,
    params: Optional[Dict[str, Any]] = None,
    body: Optional[Dict[str, Any]] = None,
) -> Tuple[Optional[Dict[str, Any]], Optional[str]]:
    url = f"{BASE_URL}/{endpoint}"
    query = {"access_token": token}
    if params:
        query.update(params)
    url = f"{url}?{urllib.parse.urlencode(query)}"

    data, error = http_request(method, url, body=body)
    if error:
        return None, error

    if data and isinstance(data, dict) and "error" in data:
        err = data.get("error", {})
        message = err.get("message", "Unknown error")
        code = err.get("code", "unknown")
        return None, f"{message} (code: {code})"

    return data, None


def log_result(permission: str, endpoint: str, success: bool, details: str) -> None:
    results.append(TestResult(permission, endpoint, success, details))
    status = "PASS" if success else "FAIL"
    print(f"{status} | {permission} | {endpoint}")
    if details:
        if success:
            print(f"       {details[:100]}")
        else:
            for line in details.splitlines():
                print(f"       {line}")


def log_skip(permission: str, endpoint: str, reason: str) -> None:
    skipped.append(TestResult(permission, endpoint, False, reason))
    print(f"SKIP | {permission} | {endpoint}")
    if reason:
        print(f"       {reason}")


# Tests

def test_pages_read_user_content(token: str, page_id: str) -> None:
    print("\n" + "=" * 60)
    print("Testing: pages_read_user_content")
    print("=" * 60)

    data, error = make_request(
        token,
        "GET",
        f"{page_id}/feed",
        {"fields": "id,message,from,created_time", "limit": 5},
    )

    if error:
        log_result("pages_read_user_content", f"{page_id}/feed", False, error)
    elif data and "data" in data:
        log_result(
            "pages_read_user_content",
            f"{page_id}/feed",
            True,
            f"Found {len(data.get('data', []))} posts",
        )
    else:
        log_result("pages_read_user_content", f"{page_id}/feed", False, "No data returned")


def test_pages_read_engagement(token: str, page_id: str) -> None:
    print("\n" + "=" * 60)
    print("Testing: pages_read_engagement")
    print("=" * 60)

    metrics = os.environ.get(
        "FB_INSIGHTS_METRICS",
        "page_engaged_users,page_post_engagements",
    )
    data, error = make_request(
        token,
        "GET",
        f"{page_id}/insights",
        {"metric": metrics, "period": "day"},
    )

    if error:
        log_result("pages_read_engagement", f"{page_id}/insights", False, error)
    elif data and "data" in data:
        log_result(
            "pages_read_engagement",
            f"{page_id}/insights",
            True,
            f"Got {len(data.get('data', []))} metrics",
        )
    else:
        log_result("pages_read_engagement", f"{page_id}/insights", False, "No data returned")

    posts_data, posts_error = make_request(
        token,
        "GET",
        f"{page_id}/posts",
        {"fields": "id,message,likes.summary(true),comments.summary(true),shares", "limit": 5},
    )

    if posts_error:
        log_result("pages_read_engagement", f"{page_id}/posts (engagement)", False, posts_error)
    elif posts_data and "data" in posts_data:
        log_result(
            "pages_read_engagement",
            f"{page_id}/posts (engagement)",
            True,
            f"Got engagement for {len(posts_data.get('data', []))} posts",
        )


def test_pages_messaging(token: str, page_id: str) -> None:
    print("\n" + "=" * 60)
    print("Testing: pages_messaging")
    print("=" * 60)

    recipient_id = os.environ.get("FB_MESSAGING_RECIPIENT_ID")
    if recipient_id:
        data, error = make_request(
            token,
            "POST",
            f"{page_id}/messages",
            {},
            {
                "recipient": {"id": recipient_id},
                "message": {"text": f"[API Test] Message sent at {datetime.utcnow().isoformat()}Z"},
            },
        )

        if error:
            log_result("pages_messaging", f"{page_id}/messages (SEND)", False, error)
        elif data and data.get("message_id"):
            log_result(
                "pages_messaging",
                f"{page_id}/messages (SEND)",
                True,
                f"Message ID: {data.get('message_id')}",
            )
        else:
            log_result("pages_messaging", f"{page_id}/messages (SEND)", False, "No message_id returned")
        return

    data, error = make_request(
        token,
        "GET",
        f"{page_id}/conversations",
        {"fields": "participants,updated_time", "limit": 1},
    )

    if error:
        log_result("pages_messaging", f"{page_id}/conversations", False, error)
    elif data and "data" in data:
        log_result(
            "pages_messaging",
            f"{page_id}/conversations",
            True,
            f"Found {len(data.get('data', []))} conversations",
        )
    else:
        log_result("pages_messaging", f"{page_id}/conversations", False, "No data returned")


def test_pages_manage_posts(token: str, page_id: str) -> None:
    print("\n" + "=" * 60)
    print("Testing: pages_manage_posts")
    print("=" * 60)

    data, error = make_request(
        token,
        "POST",
        f"{page_id}/feed",
        {},
        {"message": f"[API Test] Created at {datetime.utcnow().isoformat()}Z", "published": False},
    )

    if error:
        log_result("pages_manage_posts", f"{page_id}/feed (CREATE)", False, error)
        return

    post_id = data.get("id") if data else None
    if post_id:
        log_result("pages_manage_posts", f"{page_id}/feed (CREATE)", True, f"Created: {post_id}")
        del_data, del_error = make_request(token, "DELETE", post_id)
        if del_error:
            log_result("pages_manage_posts", f"{post_id} (DELETE)", False, del_error)
        elif del_data and del_data.get("success"):
            log_result("pages_manage_posts", f"{post_id} (DELETE)", True, "Deleted successfully")
    else:
        log_result("pages_manage_posts", f"{page_id}/feed (CREATE)", False, "No post ID returned")


def test_pages_manage_engagement(token: str, page_id: str) -> None:
    print("\n" + "=" * 60)
    print("Testing: pages_manage_engagement")
    print("=" * 60)

    posts_data, _ = make_request(
        token,
        "GET",
        f"{page_id}/posts",
        {"fields": "id,message", "limit": 1},
    )

    if not posts_data or not posts_data.get("data"):
        log_result("pages_manage_engagement", "GET posts", False, "No posts found")
        return

    post_id = posts_data["data"][0].get("id")
    if not post_id:
        log_result("pages_manage_engagement", "GET posts", False, "No post ID")
        return

    comments_data, error = make_request(
        token,
        "GET",
        f"{post_id}/comments",
        {"fields": "id,message,from,created_time", "limit": 5},
    )

    if error:
        log_result("pages_manage_engagement", f"{post_id}/comments (READ)", False, error)
    elif comments_data and "data" in comments_data:
        log_result(
            "pages_manage_engagement",
            f"{post_id}/comments (READ)",
            True,
            f"Got {len(comments_data.get('data', []))} comments",
        )

        if comments_data.get("data"):
            comment_id = comments_data["data"][0].get("id")
            if comment_id:
                reply_data, reply_error = make_request(
                    token,
                    "POST",
                    f"{comment_id}/comments",
                    {},
                    {"message": "[API Test Reply] Thank you!"},
                )

                if reply_error:
                    log_result("pages_manage_engagement", f"{comment_id}/comments (REPLY)", False, reply_error)
                elif reply_data and reply_data.get("id"):
                    log_result(
                        "pages_manage_engagement",
                        f"{comment_id}/comments (REPLY)",
                        True,
                        f"Created reply: {reply_data.get('id')}",
                    )
                    make_request(token, "DELETE", reply_data.get("id"))


def test_business_management(token: str) -> None:
    print("\n" + "=" * 60)
    print("Testing: business_management")
    print("=" * 60)

    data, error = make_request(token, "GET", "me/businesses", {"fields": "id,name,created_time"})
    if error:
        log_result("business_management", "me/businesses", False, error)
    elif data and "data" in data:
        log_result("business_management", "me/businesses", True, f"Found {len(data.get('data', []))} businesses")
        if data.get("data"):
            business_id = data["data"][0].get("id")
            if business_id:
                pages_data, pages_error = make_request(
                    token,
                    "GET",
                    f"{business_id}/owned_pages",
                    {"fields": "id,name"},
                )
                if pages_error:
                    log_result("business_management", f"{business_id}/owned_pages", False, pages_error)
                elif pages_data and "data" in pages_data:
                    log_result(
                        "business_management",
                        f"{business_id}/owned_pages",
                        True,
                        f"Found {len(pages_data.get('data', []))} pages",
                    )


def test_additional_permissions(
    user_token: Optional[str],
    page_token: str,
    page_id: str,
) -> None:
    print("\n" + "=" * 60)
    print("Testing: Additional permissions (verification)")
    print("=" * 60)

    if user_token:
        accounts_data, accounts_error = make_request(
            user_token,
            "GET",
            "me/accounts",
            {"fields": "id,name,category"},
        )

        if accounts_error:
            log_result("pages_show_list", "me/accounts", False, accounts_error)
        elif accounts_data and "data" in accounts_data:
            log_result(
                "pages_show_list",
                "me/accounts",
                True,
                f"Found {len(accounts_data.get('data', []))} pages",
            )
    else:
        log_skip("pages_show_list", "me/accounts", "Requires user access token")

    page_data, page_error = make_request(
        page_token,
        "GET",
        page_id,
        {"fields": "id,name,about,category,phone,website"},
    )

    if page_error:
        log_result("pages_manage_metadata", page_id, False, page_error)
    elif page_data and page_data.get("id"):
        log_result("pages_manage_metadata", page_id, True, f"Page: {page_data.get('name')}")


def print_summary() -> None:
    print("\n" + "=" * 60)
    print("TEST SUMMARY")
    print("=" * 60)

    passed = len([r for r in results if r.success])
    failed = len([r for r in results if not r.success])
    skipped_count = len(skipped)

    print(f"\nTotal tests: {len(results)}")
    print(f"Passed: {passed}")
    print(f"Failed: {failed}")
    if skipped_count:
        print(f"Skipped: {skipped_count}")

    if failed > 0:
        print("\nFailed tests:")
        for r in results:
            if not r.success:
                print(f"  - {r.permission}: {r.endpoint}")
                print(f"    {r.details[:80]}")
    if skipped_count:
        print("\nSkipped tests:")
        for r in skipped:
            print(f"  - {r.permission}: {r.endpoint}")
            print(f"    {r.details[:80]}")

    required_failed = sorted(
        {
            r.permission
            for r in results
            if (not r.success and r.permission in REQUIRED_PERMISSIONS)
        }
    )
    if required_failed:
        print("\nRequired permissions failing:")
        for perm in required_failed:
            print(f"  - {perm}")


def get_page_token_from_user_token(
    user_token: str,
    page_id: Optional[str],
) -> Tuple[Optional[str], Optional[str], Optional[str]]:
    data, error = make_request(
        user_token,
        "GET",
        "me/accounts",
        {"fields": "id,name,access_token,category", "limit": 50},
    )

    if error or not data or "data" not in data:
        return None, None, None

    pages = data.get("data") or []
    if not pages:
        return None, None, None

    chosen = None
    if page_id:
        for page in pages:
            if page.get("id") == page_id:
                chosen = page
                break
    if chosen is None:
        chosen = pages[0]

    return chosen.get("id"), chosen.get("access_token"), chosen.get("name")


def fetch_from_supabase() -> Tuple[str, str, str]:
    supabase_url = os.environ.get("NEXT_PUBLIC_SUPABASE_URL", "").rstrip("/")
    service_key = os.environ.get("SUPABASE_SERVICE_ROLE_KEY", "")

    if not supabase_url or not service_key:
        raise RuntimeError("Missing Supabase URL or service role key")

    url = f"{supabase_url}/rest/v1/facebook_pages?select=*&status=eq.active&limit=1"
    headers = {
        "apikey": service_key,
        "Authorization": f"Bearer {service_key}",
    }

    data, error = http_request("GET", url, headers=headers)
    if error:
        raise RuntimeError(f"Supabase error: {error}")

    if not data or not isinstance(data, list) or not data:
        raise RuntimeError("No active pages found in database")

    page = data[0]
    page_id = page.get("page_id")
    page_name = page.get("page_name", "")
    page_access_token = page.get("page_access_token")

    if not page_id or not page_access_token:
        raise RuntimeError("Missing page_id or page_access_token in DB")

    token = decrypt_token(str(page_access_token))
    return str(page_id), token, page_name


def main() -> None:
    load_env()

    print("=" * 60)
    print("FACEBOOK GRAPH API PERMISSION TEST")
    print("=" * 60)

    token = (
        os.environ.get("FB_TOKEN")
        or os.environ.get("FACEBOOK_USER_ACCESS_TOKEN")
        or os.environ.get("FACEBOOK_PAGE_ACCESS_TOKEN")
    )
    page_id = os.environ.get("FB_PAGE_ID") or os.environ.get("FACEBOOK_PAGE_ID")

    user_token: Optional[str] = None
    page_token: Optional[str] = None
    page_name: Optional[str] = None

    if token:
        resolved_page_id, resolved_page_token, resolved_page_name = get_page_token_from_user_token(token, page_id)
        if resolved_page_id and resolved_page_token:
            user_token = token
            page_token = resolved_page_token
            page_id = resolved_page_id
            page_name = resolved_page_name
        elif page_id:
            page_token = token

    if not token or not page_id or not page_token:
        print("\nNo token in env, fetching from database...")
        try:
            page_id, page_token, page_name = fetch_from_supabase()
            token = page_token
            print(f"\nUsing page: {page_name} ({page_id})")
        except Exception as exc:
            print(f"ERROR: {exc}")
            print("\nUsage:")
            print("  FB_TOKEN=<token> FB_PAGE_ID=<page_id> python scripts/test-fb-permissions.py")
            sys.exit(1)

    if not token or not page_id or not page_token:
        print("ERROR: Missing token or page ID after lookup")
        sys.exit(1)

    print(f"\nAPI Version: {GRAPH_API_VERSION}")
    print(f"Page ID: {page_id}")
    print(f"Token: {page_token[:20]}...")

    test_pages_read_user_content(page_token, page_id)
    test_pages_read_engagement(page_token, page_id)
    test_pages_messaging(page_token, page_id)
    test_pages_manage_posts(page_token, page_id)
    test_pages_manage_engagement(page_token, page_id)

    if user_token:
        test_business_management(user_token)
    else:
        log_skip("business_management", "me/businesses", "Requires user access token")

    test_additional_permissions(user_token, page_token, page_id)

    print_summary()


if __name__ == "__main__":
    try:
        main()
    except Exception as exc:
        print(f"ERROR: {exc}")
        sys.exit(1)
